add rust's crates.io support (downloads)

kstep opened the pull request.

espadrine commented on server.js:
```diff
@@ -347,6 +347,54 @@ cache(function(data, match, sendBadge, request) {
+
+      if (mode == 'd') {
+        downloads = data.crate.downloads || data.version.downloads;
+        version = data.version.num;
```
When no version is provided, `data` seems to only contain keys `crate`, `keywords`, `versions`. This code does not work.

espadrine commented on server.js:
```diff
@@ -347,6 +347,54 @@ cache(function(data, match, sendBadge, request) {
+cache(function (data, match, sendBadge, request) {
+  var mode = match[1];
+  var crate = match[2];
+  var version = match[3];
```
Can you add a comment giving examples of what mode, crate and version may hold?

espadrine commented on server.js:
```diff
@@ -347,6 +347,54 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+camp.route(/^\/crates\/(d|v|dv)\/([A-Za-z0-9_-]+)(?:\/([0-9.]+))?\.(svg|png|gif|jpg|json)$/,
```
Can you add a comment saying `// Rust download and version integration` please?

espadrine commented on try.html:
```diff
@@ -211,6 +211,10 @@
+  <tr><th> Crates.io: </th>
+  <td><img src='/crates/d/rustc-serialize.svg' alt=''/></td>
+  <td><code>https://img.shields.io/crates/d/rustc-serialize.svg</code></td>
+  </tr>
```
Add an example for `/dv/`

espadrine commented:
Thanks!

If the code with a specified version is too different from the code without one, you may have two distinct functions for it, with two different regexes.

kstep commented:
I'm not sure it's worth to split version and downloads code in this case.

kstep commented on server.js:
```diff
@@ -347,6 +347,55 @@ cache(function(data, match, sendBadge, request) {
+      } else if (mode = 'dv') {
+        downloads = data.version? data.version.downloads: data.versions[0].downloads;
+        version = data.version && data.version.num;
+        badgeData.text[1] = metric(downloads) + (version? ' version ' + version: ' latest version');
```
I'm not sure about this line. If user has chosen "dv" mode and no version, should it show latest version number, or just "latest version" phrase? What do you think?

espadrine replied:
In this case, `version` (the one from the URL) is shadowed, so it looks like that `?:` doesn't give the right answer.

kstep replied:
Actually, no. If the request is for concrete version, then data will be `{ "version": { ... } }`, so `version = data.version && data.version.num` will have the version number returned from crates.io, but if the request was for whole crate, the data will be `{ "crate": { ... }, "versions": [ ... ], ... }, and `version = data.version && ...` will be falsy.

kstep replied:
Technically it's not necessary to shadow version once again, because if the request was successful, and it was for concrete version, the `data.version.num` should be equal to original `version` from the URL, but I left it this way as an additional sanity check. Maybe I'm overprotective here.

espadrine replied:
I can't help but notice that the code is still wrong. I just checked out your branch, and the version badge gives download information. Are you sure it's better to have all of those running through the same function?

espadrine commented:
From #cargo:

> \<acrichto\> espadrine: the API isn't necessarily super stable (very little thought has gone into it), but we could leave it around if necessary (and try to iterate on v2 later)

kstep commented:
Which means?... What have you decided?

espadrine commented on try.html:
```diff
@@ -282,6 +290,10 @@
   <td><img src='/hackage/v/lens.svg' alt=''/></td>
   <td><code>https://img.shields.io/hackage/v/lens.svg</code></td>
   </tr>
+  <tr><th> Crates.io: </th>
```
Please use a `data-keywords='Rust'`, here and on the other two.

espadrine commented:
As an aside, please squash your commits together; I don't want to have partially dysfunctional commits.

Thanks for your work!

kstep commented:
OK, I fixed everything you pointed out (`/v/` mode not showing version was actually a classical beginner's mistake: `=` instead of `==`, sorry for not paying enough attention) and squashed everything into one commit.

espadrine commented:
Thanks a lot!
